File: frontend/src/types/workflow.ts
/**
 * Workflow DAG contract — mirrors ``models/workflow.py`` and
 * ``workflow.describe_blocks()``.
 *
 * ### Field names that used to be wrong
 *
 * A node carries ``type``/``label``/``params``; the old client types called the
 * same things ``block_type``/``name``/``config``, so the canvas rendered blank
 * labels and an editor that wrote ``config`` was rejected. The palette ships
 * ``params`` as a list of required parameter *names* plus ``default_params``.
 */

import { IssueSeverity } from "@/types/common";
import { ApprovalStage } from "@/types/common";

/** The kinds of block a production flow can contain. */
export type WorkflowNodeType =
  | "research"
  | "script"
  | "lint"
  | "gate"
  | "voiceover"
  | "scenes"
  | "ai_assist"
  | "timeline_check"
  | "render_plan"
  | "publish"
  | "ingest_external";

/** One draggable block in the flow canvas. */
export interface WorkflowNode {
  id: string;
  type: WorkflowNodeType;
  label: string;
  x: number;
  y: number;
  enabled: boolean;
  params: Record<string, unknown>;
}

/** A directed link between two blocks. */
export interface WorkflowEdge {
  id: string;
  source: string;
  target: string;
}

/** A production flow (`Workflow`). */
export interface Workflow {
  name: string;
  nodes: WorkflowNode[];
  edges: WorkflowEdge[];
  version: number;
  updated_at: string;
}

/** One pre-save finding. */
export interface WorkflowCheckIssue {
  code: string;
  severity: IssueSeverity;
  message: string;
  hint?: string | null;
  node_id?: string | null;
}

/** The pre-save gate's verdict (`WorkflowChecklist`). */
export interface WorkflowChecklist {
  issues: WorkflowCheckIssue[];
  ready: boolean;
  node_count: number;
  edge_count: number;
  order: string[];
}

/** Payload for ``PUT /projects/{id}/workflow`` and the checklist POST. */
export interface WorkflowSave {
  workflow: Workflow;
  force?: boolean;
}

/** Payload for ``POST /projects/{id}/workflow/run`` (`WorkflowRunRequest`). */
export interface WorkflowRunRequest {
  inputs?: Record<string, unknown>;
}

/** Lifecycle of a flow run. */
export type WorkflowRunStatus = "running" | "completed" | "failed";

/** Lifecycle of one executed block. */
export type WorkflowStepStatus = "pending" | "running" | "succeeded" | "failed" | "skipped";

/** One executed block of a run (`WorkflowStepResult`). */
export interface WorkflowStepResult {
  node_id: string;
  label: string;
  type: WorkflowNodeType;
  status: WorkflowStepStatus;
  duration_ms: number;
  input: Record<string, unknown>;
  output: Record<string, unknown>;
  error?: string | null;
}

/** A flow run (`WorkflowRun`). */
export interface WorkflowRun {
  id: string;
  project_id: string;
  status: WorkflowRunStatus;
  inputs: Record<string, unknown>;
  outputs: Record<string, unknown>;
  steps: WorkflowStepResult[];
  message?: string | null;
  started_at: string;
  finished_at?: string | null;
  duration_ms: number;
}

/**
 * One entry of the block palette (``GET /workflow/blocks``).
 *
 * ``params`` is the list of *required* parameter names for the block; the values
 * they start with live in ``default_params``.
 */
export interface WorkflowBlockDef {
  type: WorkflowNodeType;
  label: string;
  description: string;
  params: string[];
  default_params: Record<string, unknown>;
  icon: string;
}

/** A gate block's parameters. */
export interface GateParams {
  stage: ApprovalStage;
}

File: src/content_factory/models/workflow.py
"""Node/edge workflow definition and run bookkeeping."""

from __future__ import annotations

import enum
from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field

from .common import (
    IssueSeverity,
    utcnow,
)


class WorkflowNodeType(enum.StrEnum):
    """The kinds of block a production flow can contain."""

    RESEARCH = "research"
    SCRIPT = "script"
    LINT = "lint"
    GATE = "gate"
    VOICEOVER = "voiceover"
    SCENES = "scenes"
    AI_ASSIST = "ai_assist"
    TIMELINE_CHECK = "timeline_check"
    RENDER_PLAN = "render_plan"
    PUBLISH = "publish"
    INGEST_EXTERNAL = "ingest_external"


class WorkflowNode(BaseModel):
    """One draggable block in the flow canvas."""

    id: str
    type: WorkflowNodeType
    label: str = ""
    x: float = 0.0
    y: float = 0.0
    enabled: bool = True
    params: dict[str, Any] = Field(default_factory=dict)


class WorkflowEdge(BaseModel):
    """A link between two blocks: the target consumes the source's output."""

    id: str
    source: str
    target: str


class Workflow(BaseModel):
    """A drag-and-drop production flow attached to one project."""

    name: str = "Production pipeline"
    nodes: list[WorkflowNode] = Field(default_factory=list)
    edges: list[WorkflowEdge] = Field(default_factory=list)
    version: int = Field(default=1, ge=1)
    updated_at: datetime = Field(default_factory=utcnow)


class WorkflowCheckIssue(BaseModel):
    """One line of the pre-save checklist."""

    code: str
    severity: IssueSeverity
    message: str
    hint: str | None = None
    node_id: str | None = None


class WorkflowChecklist(BaseModel):
    """The pre-save checklist plus the resolved execution order."""

    issues: list[WorkflowCheckIssue] = Field(default_factory=list)
    ready: bool = True
    node_count: int = 0
    edge_count: int = 0
    order: list[str] = Field(default_factory=list)


class WorkflowSave(BaseModel):
    """Payload for saving a flow; ``force`` skips a failing checklist."""

    workflow: Workflow
    force: bool = False


class WorkflowRunStatus(enum.StrEnum):
    """Outcome of a flow run."""

    RUNNING = "running"
    OK = "ok"
    FAILED = "failed"
    BLOCKED = "blocked"


class WorkflowStepStatus(enum.StrEnum):
    """Outcome of a single block inside a run."""

    PENDING = "pending"
    RUNNING = "running"
    OK = "ok"
    SKIPPED = "skipped"
    FAILED = "failed"
    BLOCKED = "blocked"


class WorkflowStepResult(BaseModel):
    """What one block did, for the monitoring panel."""

    node_id: str
    label: str
    type: WorkflowNodeType
    status: WorkflowStepStatus = WorkflowStepStatus.PENDING
    duration_ms: int = 0
    input: dict[str, Any] = Field(default_factory=dict)
    output: dict[str, Any] = Field(default_factory=dict)
    error: str | None = None


class WorkflowRun(BaseModel):
    """A full run record: input variables, per-block detail, and timing."""

    id: str
    project_id: str
    status: WorkflowRunStatus = WorkflowRunStatus.RUNNING
    inputs: dict[str, Any] = Field(default_factory=dict)
    outputs: dict[str, Any] = Field(default_factory=dict)
    steps: list[WorkflowStepResult] = Field(default_factory=list)
    message: str | None = None
    started_at: datetime = Field(default_factory=utcnow)
    finished_at: datetime | None = None
    duration_ms: int = 0


class WorkflowRunRequest(BaseModel):
    """Payload for running a flow (``inputs`` are the input variables)."""

    inputs: dict[str, Any] = Field(default_factory=dict)
